contracts/InstantTrade.sol

```diff
@@ -2,66 +2,208 @@ pragma solidity ^0.4.13;
 
 import "./TokenStore.sol";
 
+contract ZeroExchange {
+    
+  address public TOKEN_TRANSFER_PROXY_CONTRACT;
+    
+  function fillOrKillOrder(
+    address[5] orderAddresses,
+    uint[6] orderValues,
+    uint fillTakerTokenAmount,
+    uint8 v,
+    bytes32 r,
+    bytes32 s)
+    public {}
+        
+  /*  
+  function fillOrder(
+    address[5] orderAddresses,
+    uint[6] orderValues,
+    uint fillTakerTokenAmount,
+    bool shouldThrowOnInsufficientBalanceOrAllowance,
+    uint8 v,
+    bytes32 r,
+    bytes32 s)
+    public
+    returns (uint filledTakerTokenAmount)
+    {}
+    */
+}
+
+contract WETH {
+  function deposit() public payable {}
+  function withdraw(uint wad) public {}
+  function balanceOf(address _owner) public constant returns (uint256 balance) {}
+  function approve(address _spender, uint256 _value) public returns (bool success) {}
+  function transfer(address _to, uint256 _value) public returns (bool success) {}
+  function transferFrom(address _from, address _to, uint256 _value) public returns (bool success) {}
+}
+
 contract InstantTrade is SafeMath, Ownable {
 
-  // This is needed so we can withdraw funds from other smart contracts
-  function() payable {
+  address public wETH;
+  address public zeroX;
+  address public proxyZeroX;
+    
+  mapping(address => bool) allowedFallbacks; // Limit fallback to avoid accidental ETH transfers
+    
+  constructor(address _weth, address _zeroX) Ownable() public {
+    wETH = _weth;
+    zeroX = _zeroX;
+    proxyZeroX = ZeroExchange(zeroX).TOKEN_TRANSFER_PROXY_CONTRACT();
+       
+    allowedFallbacks[wETH] = true;
+  }
+   
+  // Only allow incoming ETH from known contracts (Exchange and WETH withdrawals)
+  function() public payable {
+    require(allowedFallbacks[msg.sender]);
+  }
+  
+  
+  function allowFallback(address _contract, bool _allowed) external onlyOwner {
+    allowedFallbacks[_contract] = _allowed;
   }
   
-  // End to end trading in a single call
+  
+  // End to end trading in a single call (Token Store, EtherDelta)
   function instantTrade(address _tokenGet, uint _amountGet, address _tokenGive, uint _amountGive,
-      uint _expires, uint _nonce, address _user, uint8 _v, bytes32 _r, bytes32 _s, uint _amount, address _store) payable {
+    uint _expires, uint _nonce, address _user, uint8 _v, bytes32 _r, bytes32 _s, uint _amount, address _store) external payable {
     
     // Fix max fee (0.4%) and always reserve it
     uint totalValue = safeMul(_amount, 1004) / 1000;
```

Comment:
Since we started to reuse it, can we move the fee into a contract const? Fine to make it not changeable, but let's put it somewhere in a single place